optaplanner-core/src/main/java/org/optaplanner/core/impl/constructionheuristic/decider/MultiThreadedConstructionHeuristicDecider.java

```diff
@@ -111,12 +111,14 @@ public void phaseStarted(ConstructionHeuristicPhaseScope<Solution_> phaseScope)
     @Override
     public void phaseEnded(ConstructionHeuristicPhaseScope<Solution_> phaseScope) {
         super.phaseEnded(phaseScope);
+        // Clear non-started operations to improve shutdown performance. We're shutting down anyway, so it's fine.
+        operationQueue.clear();
```

Comment:
Just thinking of this now:
Can this cause a race condition?
- The ApplyStep operations have a `CyclicBarrier#await()` in MoveThreadRunner line 125.
- Presume there are 4 child threads
- At some piont, we do a step, adding 4 ApplyStep operations on the operationsQueue.
- 3 ApplyStep operations gets pulled quickly, but the 4th child thread is still working on a very big move form the previous step (which will be discarded when it finishes).
- Those first 3 ApplyStep operations hit the `moveThreadBarrier.await();`, waiting for the 4th one to come along.
- Meanwhile, time's up. Termination sets in. This `operationQueue.clear();` activates and 4 DestroyOperations are added.
- When the 4th child thread finishes his big move, its ApplyStep operation is already cleared. It pulls a DestroyOperation and kills itself.
- The other 3 child threads hang on the `moveThreadBarrier.await();` forever. This is a bug. Later, the thread interruptions kill the child threads. There is no destroy operations, so no info is given back, including an accurate "score calculation speed" etc.

Reply:
This wasn't resolved and is proven to be accurate.